lib/Feature/FeatureModelParser.cpp

```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+  }
+
+  if (MinCardinality.value() != 1 || (MaxCardinality.value() != 1 && MaxCardinality.value() != UINT_MAX)) {
+    std::cerr << "Cardinality unsupported. We support cardinalities [1,1] (alternative) or [1, *] (or group)." << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+
+  return std::optional<std::tuple<int, int>>(std::tuple<int, int>{MinCardinality.value(), MaxCardinality.value()});
+}
+
+
+/// This method parses the given cardinality and returns an optional.
+/// If the optional is empty, the process failed; otherwise the result contains
+/// either UINT_MAX for the wildcard or the cardinality number as integer.
+/// \param CardinalityString the cardinality to parse
+/// \return an optional that contains no integer in case of failure or UINT_MAX
+/// for wildcard, or the number itself.
+std::optional<int> FeatureModelSxfmParser::parseCardinality(const string& CardinalityString) {
+  std::optional<int> Result = std::optional<int>();
+  if (CardinalityString == "*") {
+    // We use UINT_MAX as our magic integer (which is -1 as int) to indicate
+    // that the cardinality is a wildcard.
+    Result = UINT_MAX;
+  } else {
+    // Convert the string into an integer in a safe way
+    char *End;
+    long LongNumber;
+    errno = 0;
+    LongNumber = strtol(CardinalityString.c_str(), &End, 0);
+    if (errno == ERANGE || LongNumber < INT_MIN || LongNumber > INT_MAX
+        || *CardinalityString.c_str() == '\0' || *End != '\0') {
+      std::cerr << "The following cardinality is not integer: " << CardinalityString << std::endl;
+    } else {
+      Result = LongNumber;
+    }
+  }
+
+  return Result;
+}
+
+/// This method reads beginning from a certain starting position and reads in
+/// every character until CharToSearch is found.
+/// \param StringToReadFrom string to read from
+/// \param CharToSearch character to search for
+/// \param Start the starting position to start reading
+/// \return the string read until the given character is found
+string FeatureModelSxfmParser::readUntil(const string& StringToReadFrom, const char& CharToSearch, std::string::size_type Start) {
+  string Result;
+  std::string::size_type Pos = Start;
+  if (Pos > StringToReadFrom.length()) {
+    std::cerr << "Position out of bounds!" << std::endl;
+    return "";
+  }
+
+  while ((Pos < StringToReadFrom.length()) && (StringToReadFrom.at(Pos) != CharToSearch)) {
+    Result += StringToReadFrom.at(Pos);
+    Pos++;
+  }
+  return Result;
+}
+
+/// This method counts the occurrences of the second argument in the first argument.
+/// The intended use of this method is to count the indentation level of the feature
+/// tree in the sxfm format.
+/// \param StringToSearch the string to search in
+/// \param StringToFind the string to count the occurrences of
+/// \return the number of occurrences
+int FeatureModelSxfmParser::countOccurrences(const string& StringToSearch, const string& StringToFind) {
```

Comment:
```suggestion
int FeatureModelSxfmParser::countOccurrences(const string& StringToSearch, const string& StringToFind) {
```
Has different parameter names in the header file.